### PR TITLE
feat: add Breez SDK wallet

### PR DESCRIPTION
**Pull request by callebtc**
Set config in `.env`:

```
LNBITS_BACKEND_WALLET_CLASS=BreezSdkWallet

BREEZ_GREENLIGHT_SEED = "..."
BREEZ_API_KEY = "..."
BREEZ_GREENLIGHT_INVITE_CODE = "..."
```

### CODE REVIEW

**Comment by dni** [comment]
wouldnt be the greenlight client enough for the fundingsource? https://github.com/Blockstream/greenlight

it has this sweet python package aswell https://pypi.org/project/gl-client/
```poetry add gl-client```

**Comment by arcbtc** [comment]
Im with DNI I think, we should dump breezesdk support, as its greenlight with bunch of lnbits features on top. So why not just have greenlight

**Comment by arcbtc** [comment]
Stance reverted. I get it now, is great for a company using breez sdk to be able to run lnbits on top

**Comment by bitkarrot** [comment]
Greenlight Please. I will use it!

**Comment by prusnak** [comment]
I think we should postpone this PR until there is a proper PyPI Breeze-SDK package for Python.

Blocked by https://github.com/breez/breez-sdk/issues/332

**Comment by kingonly** [comment]
https://github.com/breez/breez-sdk/issues/332 was fixed:
https://pypi.org/project/breez-sdk-lib/0.2.9/

**Comment by prusnak** [comment]
> https://github.com/breez/breez-sdk/issues/332 was fixed:
> 
> https://pypi.org/project/breez-sdk-lib/0.2.9/

Great! Thanks for the update! 🎉

**Comment by prusnak** [comment]
1. rebased the PR on top of current `dev` branch.
2. commit f6d6fe64b91293ce9fc9686cfd0d0dae100b160d Removed no longer needed `lnbits/wallets/breez_sdk_files` and used `breez_sdk_lib` from PyPI instead.

**Comment by prusnak** [comment]
1. rebased the PR on top of current `dev` branch.
2. commit e31018a397acdba7718d8b184b218bf0db65d6fd removed no longer needed `lnbits/wallets/breez_sdk_files` and used `breez_sdk_lib` from PyPI instead.

File: .env.example
Comment (kingonly) [reply .env.example:84]: in the Breez SDK the invite code and GL certificate are interchangeable

Comment (prusnak) [reply .env.example:85]: added comment in https://github.com/lnbits/lnbits/pull/1897/commits/4b335f963fa9a04274241e145fd62fbe51dcd207

**Comment by prusnak** [comment]
Squashed and rebased on top of the current `dev`.

File: lnbits/wallets/breez.py
Comment (aljazceru) [inline lnbits/wallets/breez.py:95]: breez expects msats but in UI its sats so the amount will be off by 1000

Comment (prusnak) [reply lnbits/wallets/breez.py:95]: Ah, that explains the issue I saw when I was testing it!

Comment (aljazceru) [reply lnbits/wallets/breez.py:95]: same problem on the wallet balance, just the other way around - breez returns msats so the balance in lnbits is off by x1000

or maybe i've just discovered a bitcoin faucet :D

Comment (prusnak) [reply lnbits/wallets/breez.py]: Fixed in 9c54b11241f89527342ef5d68a419c40a18cad7c

**Comment by prusnak** [comment]
As of commit ac2987c one can:
1. receive sats
2. try to send sats, but the attempt fails with the following error:

```
HTTPException 520: Payment failed: payment failed: SendPaymentError.PaymentTimeout('Payment timeout: status: Unknown, message: "Error calling method Pay: RpcError { code: Some(210), message: \\"Ran out of routes to try after 61 attempts: see `paystatus`\\", data: None }", details: [], metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Fri, 22 Dec 2023 22:25:31 GMT", "content-length": "0"} }')
```

However, this seems like a liquidity issue of a remote node (I tried several nodes, none worked).

Another issue is that the SDK creates two files (`storage.sql` and `sync_storage.sql`) in the current directory. This is ugly and will fail on Nix for example.

I will ask if there is a way how to override the working dir -> https://github.com/breez/breez-sdk/issues/701

**Comment by kingonly** [comment]
Are you using an old or a new channel?
Where are you trying to send the funds?
Integrating this would be helpful: https://sdk-doc.breez.technology/guide/failure_report.html

**Comment by prusnak** [comment]
> Are you using an old or a new channel?

I have no idea. I am just using the invite codes I was given.

> Where are you trying to send the funds?

To Wallet of Satoshi

> Integrating this would be helpful: https://sdk-doc.breez.technology/guide/failure_report.html

Great! I integrated it just now. Here is the payment hash which was (hopefully) reported: `9752ba842427a92cf1a9d10aa24042997b0dc47db357d5d3776125fb743e9240`

**Comment by kingonly** [comment]
Old channels, created before October 15th are non-anchor channels and won't work in a high fee environment.

File: lnbits/wallets/breez.py
Comment (prusnak) [reply lnbits/wallets/breez.py]: > same problem on the wallet balance, just the other way around - breez returns msats so the balance in lnbits is off by x1000

I don't see this issue. When I set 2501 sats to LNbits, it showed me 2501 sats.

**Comment by aljazceru** [comment]
did some more testing - payment statuses are not handled ok. for example:
paying an invoice from lnbits returns failed and keeps the payment dialog open 
image
but the invoice is actually paid 
image

File: lnbits/wallets/breez.py
Comment (dangeross) [inline lnbits/wallets/breez.py:142]: The Payment `pending` field has been replaced with Payment `status`. As the FAILED case is handled in the exception catch, it can be either PENDING or COMPLETE. 
```suggestion
            assert payment.status == breez_sdk.PaymentStatus.COMPLETE, "payment is pending"
```

File: lnbits/wallets/breez.py
Comment (dangeross) [inline lnbits/wallets/breez.py:109]: I believe here you could add setting the description hash the same way as corelightning.
```suggestion
            if description_hash and not unhashed_description:
                raise Unsupported(
                    "'description_hash' unsupported by Greenlight, provide"
                    " 'unhashed_description'"
                )
            breez_invoice: breez_sdk.ReceivePaymentResponse = (
                self.sdk_services.receive_payment(
                    breez_sdk.ReceivePaymentRequest(
                        amount * 1000,  # breez uses msat
                        unhashed_description.decode() if unhashed_description else memo,
                        preimage=kwargs.get("preimage"),
                        opening_fee_params=None,
                        use_description_hash=True if unhashed_description else None,
                    )
                )
            )
```

File: lnbits/wallets/breez.py
Comment (dangeross) [inline lnbits/wallets/breez.py:159]: You can check the payment status here
```suggestion
            payment = self.sdk_services.payment_by_hash(checking_id)
            if payment is None:
                return PaymentStatus(None)
            assert payment.payment_type == breez_sdk.PaymentType.RECEIVED
            return PaymentStatus(payment.status == breez_sdk.PaymentStatus.COMPLETE)
```

File: lnbits/wallets/breez.py
Comment (dangeross) [inline lnbits/wallets/breez.py:171]: We can additionally check if the payment has FAILED and instead of checking it's pending, check if the status is COMPLETE.
```suggestion
        async def get_payment_status(self, checking_id: str) -> PaymentStatus:
            payment = self.sdk_services.payment_by_hash(checking_id)
            if payment is None:
                return PaymentStatus(None)
            assert payment.payment_type == breez_sdk.PaymentType.SENT
            if payment.status == breez_sdk.PaymentStatus.COMPLETE:
                return PaymentStatus(
                    True, payment.fee_msat, payment.details.data.payment_preimage
                )
            elif payment.status == breez_sdk.PaymentStatus.FAILED:
                return PaymentStatus(False)
            else:
                return PaymentStatus(None)
```

File: pyproject.toml
Comment (dangeross) [inline pyproject.toml:54]: Can be bumped to 0.2.15
```suggestion
breez-sdk = {version = "0.2.15", optional = true}
```

File: lnbits/wallets/breez.py
Comment (dangeross) [inline lnbits/wallets/breez.py:58]: This could be an optional setting, as the Greenlight node could exist already and used in other applications. You could instead handle `breez_sdk.connect()` throwing an error is Greenlight rejects the connection.

File: lnbits/wallets/breez.py
Comment (dangeross) [inline lnbits/wallets/breez.py:87]: Better to check the `channels_balance_msat`, which is a sum of the total spendable amount.
```suggestion
            return StatusResponse(None, int(node_info.channels_balance_msat))
```

**Review by dangeross — COMMENTED** [review]
Hi, @dangeross from the Breez team here. Hopefully added some comments to help out with some interface changes to the Breez SDK.

File: lnbits/wallets/breez.py
Comment (prusnak) [reply lnbits/wallets/breez.py:58]: The docu says the following:

```
a GreenLight certificate can be used directly instead of a Breez/GreenLight invite code)
```

So if it is already used and running, you should specify a cert here.

Maybe we should rename the field from `BREEZ_GREENLIGHT_INVITE_CODE` to `BREEZ_GREENLIGHT_INVITE_CODE_OR_CERT` or something like that.

**Comment by prusnak** [comment]
> Hi, @dangeross from the Breez team here. Hopefully added some comments to help out with some interface changes to the Breez SDK.

Applied suggestions from @dangeross and rebased on top of current dev.

File: lnbits/wallets/breez.py
Comment (dangeross) [reply lnbits/wallets/breez.py:59]: If the node is already used and running, neither are technically needed as they are unused. To use the certificate we need to pass a slightly different GreenlightNodeConfig for the cert:
```
breez_sdk.GreenlightNodeConfig(
    partner_credentials=breez_sdk.GreenlightCredentials(
        device_key=BREEZ_GREENLIGHT_DEVICE_KEY_BYTES,
        device_cert=BREEZ_GREENLIGHT_DEVICE_CERT_BYTES,
    ),
    invite_code=None,
)
```

Maybe we can create a follow up PR to add this?

Comment (prusnak) [reply lnbits/wallets/breez.py:59]: > Maybe we can create a follow up PR to add this?

Feel free to create a new PR that targets the `wallets/breez` branch - once merged, the commits will be added to this PR.

Comment (dangeross) [reply lnbits/wallets/breez.py:59]: Here you go: PR https://github.com/lnbits/lnbits/pull/2243

**Comment by motorina0** [comment]
@callebtc is there a docker image that we can use for `regtests`?

**Comment by aljazceru** [comment]
@dangeross whats the best way to access the instance created by lnbits with breez_sdk cli?

**Comment by dangeross** [comment]
@aljazceru in the `BreezSdkWallet` class you can access the instance with `self.sdk_services`

**Comment by dangeross** [comment]
@aljazceru 

In the CLI directory of the breez-sdk repo, create a directory e.g. lnbits. In the directory create a file `phrase` and paste in the same mnemonic. 
- Run `cargo run -- --data_dir lnbits`
- Then `set_api_key <api key>`
- `connect`

**Comment by aljazceru** [comment]
> @aljazceru
> 
> In the CLI directory of the breez-sdk repo, create a directory e.g. lnbits. In the directory create a file named `phrase` and paste in the same mnemonic as used with lnbits.
> 
>     * Run `cargo run -- --data_dir lnbits`
> 
>     * Then `set_api_key <api key>`
> 
>     * `connect`

thats what i was looking for, thnx!

**Comment by dangeross** [comment]
@prusnak I've created a PR #2329 to bump the breez-sdk dependency to 0.3.6

**Comment by prusnak** [comment]
> @prusnak I've created a PR #2329 to bump the breez-sdk dependency to 0.3.6

Merged

**Comment by prusnak** [comment]
Rebased to resolve conflicts and force pushed.

**Review by prusnak — APPROVED** [review]

**Review by dni — CHANGES_REQUESTED** [review]
the LSP fees isnt substracted from incoming payments

**Comment by dangeross** [comment]
> the LSP fees isnt substracted from incoming payments

It should be subtracted, if there is no channel (or not enough inbound liquidity) to handle the payment, the LSP is fee is subtracted from the amount to be received. 

For example with no liquidity, a merchant wants to receive 10000 sats and a calculated fee of 2500 sats is applied for the LSP and when the payment is completed the merchant receives 7500 sats.

The application of fees can be pre-checked for an amount by calling:
```
req = breez_sdk.OpenChannelFeeRequest(amount_msat)
channel_fees = sdk_services.open_channel_fee(req)
```
https://sdk-doc.breez.technology/guide/receive_payment.html#calculating-fees

**Comment by aljazceru** [comment]
There is still the issue of payment status updates - lnbits first reports the payment failed and then upon closing the dialog and refreshing it shows it as succeeded.
payment:
false-failed

after refresh: 
image

Full debug log of this attempt attached:

[failed-successful-payment-debug.txt](https://github.com/lnbits/lnbits/files/15106892/failed-successful-payment-debug.txt)

**Comment by dangeross** [comment]
I've created PR https://github.com/lnbits/lnbits/pull/2467 to fix this

**Comment by dangeross** [comment]
Did that fix the problem for you @aljazceru?

**Comment by dangeross** [comment]
I've added a couple of PRs to rebase and bump SDK versions
- Rebase to latest dev: https://github.com/lnbits/lnbits/pull/2583
- Upgrade to Breez SDK v0.5.0: https://github.com/lnbits/lnbits/pull/2584

**Comment by dangeross** [comment]
PR to update Breez SDK to v0.5.0: https://github.com/lnbits/lnbits/pull/2585

**Comment by prusnak** [comment]
> PR to update Breez SDK to v0.5.0: #2585

Applied into rebased commit [495e9eb](https://github.com/lnbits/lnbits/pull/1897/commits/495e9eb00e866c075bdb598dae99172c6968e888)

**Comment by aljazceru** [comment]
@dangeross i think the update do 0.5.0 broke this branch

problem starts with error below which i'm guessing is the new addition of how connect is called on [line 133](https://github.com/lnbits/lnbits/pull/1897/files#diff-af5bc71a32b467121c4c381c6837860af2f41f29c2bd6a358846f9eec37a7b3aR133)
```
Error initializing BreezSdkWallet: cannot initialize BreezSdkWallet: connect() takes 2 positional arguments but 3 were given
```
fixing that brings us to seond problem which i didn't manage to figure it out, but in any case the current branch doesn't work 
```
ERROR | Error initializing BreezSdkWallet: cannot initialize BreezSdkWallet: 'Config' object has no attribute 'config'
```

**Comment by dangeross** [comment]
https://github.com/lnbits/lnbits/pull/2626

**Comment by aljazceru** [comment]
> ` connect_request = breez_sdk.ConnectRequest(self.config, seed)`

indeed it does, thnx

**Comment by prusnak** [comment]
> #2626

Merged into this PR

**Comment by prusnak** [comment]
Rebased on top of current dev

**Review by dni — APPROVED** [review]
i could not test it, code looks ok

**Review by prusnak — APPROVED** [review]

File: lnbits/wallets/breez.py
Comment (motorina0) [inline lnbits/wallets/breez.py:159]: add `try-except` and return `return InvoiceResponse(False, None, None, str(e))` on `except`

File: lnbits/wallets/breez.py
Comment (motorina0) [inline lnbits/wallets/breez.py:216]: Do not raise, but return `PaymentResponse`.

```suggestion
            return PaymentResponse(False, None, None, None, "payment is pending" )
```

File: lnbits/wallets/breez.py
Comment (motorina0) [inline lnbits/wallets/breez.py:228]: - add `try-except` (for rpc errors, etc) and return `PaymentPendingStatus()` on `except`
- do not raise, but instead log and `return PaymentPendingStatus()`
- `PaymentType.FAILED` should be checked also

File: lnbits/wallets/breez.py
Comment (motorina0) [inline lnbits/wallets/breez.py:238]: - add `try-except` (for rpc errors, etc) and return `PaymentPendingStatus()` on `except`
- do not raise, but instead log and `return PaymentPendingStatus()`

**Review by motorina0 — CHANGES_REQUESTED** [review]

**Comment by motorina0** [comment]
LGTM. Added some small code changes in order to follow the behaviour of the other Funding Sources.

**Pull request merged**
